File: DelSole.CrossAnalyzers.CSharp/DelSole.CrossAnalyzers.CSharp/MakeRelayCommandRefactoring.cs
using System;
using System.Composition;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CodeRefactorings;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Rename;
using Microsoft.CodeAnalysis.Text;
using Microsoft.CodeAnalysis.Simplification;
using Microsoft.CodeAnalysis.Formatting;

namespace DelSole.CrossAnalyzers.CSharp
{
    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(MakeRelayCommandRefactoring)), Shared]
    public class MakeRelayCommandRefactoring : CodeRefactoringProvider
    {
        private string Title = "Make RelayCommand<T> class";
        public async sealed override Task
          ComputeRefactoringsAsync(CodeRefactoringContext context)
        {
            var root = await context.Document.GetSyntaxRootAsync(context.
              CancellationToken).
                ConfigureAwait(false);
            // Find the node at the selection.
            var node = root.FindNode(context.Span);
            // Only offer a refactoring if the selected node is
            // a class statement node.
            var classDecl = node as ClassDeclarationSyntax;
            if (classDecl == null)
            {
                return;
            }
            var action = CodeAction.Create(title: Title,
              createChangedDocument: c =>
              MakeRelayCommandAsync(context.Document,
              classDecl, c), equivalenceKey: Title);
            // Register this code action.
            context.RegisterRefactoring(action);
        }
        private async Task<Document>
          MakeRelayCommandAsync(Document document,
          ClassDeclarationSyntax classDeclaration, CancellationToken cancellationToken)
        {
            // The class definition represented as source text
            string newImplementation = @"
class RelayCommand<T> : ICommand
{
  readonly Action<T> _execute = null;
  readonly Predicate<T> _canExecute = null;
  public RelayCommand(Action<T> execute)
    : this(execute, null)
  {
  }
  public RelayCommand(Action<T> execute, Predicate<T> canExecute)
  {
    if (execute == null)
      throw new ArgumentNullException(""execute"");
      _execute = execute;
        _canExecute = canExecute;
  }
  [System.Diagnostics.DebuggerStepThrough]
  public bool CanExecute(object parameter)
  {
    return _canExecute == null ? true : _canExecute((T)parameter);
  }
  public event EventHandler CanExecuteChanged;
  public void RaiseCanExecuteChanged()
  {
    var handler = CanExecuteChanged;
    if (handler != null)
    {
      handler(this, EventArgs.Empty);
    }
  }
  public void Execute(object parameter)
  {
    _execute((T)parameter);
  }
}
";
            // 1. ParseSyntaxTree() gets a new SyntaxTree from the source text
            // 2. GetRoot() gets the root node of the tree
            // 3. OfType<ClassDeclarationSyntax>().FirstOrDefault() retrieves the only class
            //      definition in the tree
            // 4. WithAdditionalAnnotations() is invoked for code formatting
            var newClassNode = SyntaxFactory.ParseSyntaxTree(newImplementation).
              GetRoot().DescendantNodes().
              OfType<ClassDeclarationSyntax>().
              FirstOrDefault().
              WithAdditionalAnnotations(Formatter.Annotation, Simplifier.Annotation);
            // Get the root SyntaxNode of the document
            var root = await document.GetSyntaxRootAsync(cancellationToken);
            // Generate a new CompilationUnitSyntax (which represents a code file)
            // replacing the old class with the new one
            CompilationUnitSyntax newRoot = (CompilationUnitSyntax)root.
              ReplaceNode(classDeclaration,
              newClassNode).NormalizeWhitespace();
            if ((newRoot.Usings.Any(u => u.Name.ToFullString() == "System.Windows.Input"))
              == false)
            {
                newRoot = newRoot.AddUsings(SyntaxFactory.UsingDirective(SyntaxFactory.
                  QualifiedName(SyntaxFactory.IdentifierName("System"),
                    SyntaxFactory.IdentifierName("Windows.Input"))));
            }
            // Generate a new document based on the new SyntaxNode
            var newDocument = document.WithSyntaxRoot(newRoot);
            // Return the new document
            return newDocument;
        }

    }
}

File: DelSole.CrossAnalyzers.VisualBasic/DelSole.CrossAnalyzers.VisualBasic/MakeRelayCommandRefactoring.vb
Imports System.Composition
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.CodeActions
Imports Microsoft.CodeAnalysis.CodeRefactorings
Imports Microsoft.CodeAnalysis.Formatting
Imports Microsoft.CodeAnalysis.Rename
Imports Microsoft.CodeAnalysis.Simplification
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

<ExportCodeRefactoringProvider(LanguageNames.VisualBasic, Name:=NameOf(MakeRelayCommandRefactoring)), [Shared]>
Friend Class MakeRelayCommandRefactoring
    Inherits CodeRefactoringProvider

    Public NotOverridable Overrides Async Function _
         ComputeRefactoringsAsync(context As CodeRefactoringContext) As Task
        Dim root = Await context.Document.
          GetSyntaxRootAsync(context.CancellationToken).
          ConfigureAwait(False)
        ' Find the node at the selection.
        Dim node = root.FindNode(context.Span)
        ' Only offer a refactoring if the selected node is a class statement node.
        Dim classDecl = TryCast(node, ClassStatementSyntax)
        If classDecl Is Nothing Then Return
        Dim action = CodeAction.Create("Make RelayCommand(Of T) class",
                                       Function(c) _
                                       MakeRelayCommandAsync(context.Document,
                                                             classDecl, c))
        ' Register this code action.
        context.RegisterRefactoring(action)
    End Function
    Private Async Function MakeRelayCommandAsync(document As Document,
     classDeclaration As ClassStatementSyntax, cancellationToken As CancellationToken) _
     As Task(Of Document)
        ' The class definition represented as source text
        Dim newImplementation = "Class RelayCommand(Of T)
   Implements ICommand
   Private ReadOnly _execute As Action(Of T)
   Private ReadOnly _canExecute As Predicate(Of T)
   Public Sub New(ByVal execute As Action(Of T))
     Me.New(execute, Nothing)
   End Sub
   Public Sub New(ByVal execute As Action(Of T), ByVal canExecute As Predicate(Of T))
     If execute Is Nothing Then
       Throw New ArgumentNullException(""execute"")
     End If
     _execute = execute
     _canExecute = canExecute
   End Sub
   <DebuggerStepThrough> _
   Public Function CanExecute(ByVal parameter As Object) As Boolean Implements ICommand.CanExecute
     Return If(_canExecute Is Nothing, True, _canExecute(CType(parameter, T)))
   End Function
   Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged
   Public Sub RaiseCanExecuteChanged()
     RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
   End Sub
   Public Sub Execute(ByVal parameter As Object) Implements ICommand.Execute
     _execute(CType(parameter, T))
   End Sub
 End Class"
        ' 1. ParseSyntaxTree() gets a new SyntaxTree from the source text
        ' 2. GetRoot() gets the root node of the tree
        ' 3. OfType(Of ClassDeclarationSyntax)().FirstOrDefault()
        '    retrieves the only class definition in the tree
        ' 4. WithAdditionalAnnotations() Is invoked for code formatting
        Dim newClassNode = SyntaxFactory.ParseSyntaxTree(newImplementation).
        GetRoot().DescendantNodes().
          OfType(Of ClassBlockSyntax)().
          FirstOrDefault().
          WithAdditionalAnnotations(Formatter.Annotation, Simplifier.Annotation)
        Dim parentBlock = CType(classDeclaration.Parent, ClassBlockSyntax)
        Dim root = Await document.GetSyntaxRootAsync(cancellationToken)
        ' Generate a new CompilationUnitSyntax (which represents a code file)
        ' replacing the old class with the new one
        Dim newRoot As CompilationUnitSyntax =
          root.ReplaceNode(parentBlock, newClassNode)
        'Detect if an Imports System.Windows.Input directive already exists
        If Not newRoot.Imports.Any(Function(i) i.ToFullString.
          Contains("System.Windows.Input")) Then
            'If not, add one
            Dim newImp = SyntaxFactory.
              ImportsStatement(SyntaxFactory.
              SingletonSeparatedList(Of ImportsClauseSyntax) _
              (SyntaxFactory.
              SimpleImportsClause(SyntaxFactory.
              ParseName("System.Windows.Input"))))
            newRoot = newRoot.AddImports(newImp)
        End If
        Dim newDocument = document.WithSyntaxRoot(newRoot)
        Return newDocument
    End Function

End Class